src/mail-app/mail/model/MailListModel.ts

```diff
@@ -0,0 +1,314 @@
+ */
+export class MailListModel {
+	// Id = MailSetEntry element id
+	private readonly listModel: ListModel<LoadedMail, Id>
+
+	// keep a reverse map for going from Mail element id -> LoadedMail
+	private readonly mailMap: Map<Id, LoadedMail> = new Map()
+
+	constructor(
+		private readonly mailSet: MailFolder,
+		private readonly conversationPrefProvider: ConversationPrefProvider,
+		private readonly entityClient: EntityClient,
+	) {
+		this.listModel = new ListModel({
+			fetch: async (lastFetchedItem, count) => {
+				const lastFetchedId = lastFetchedItem?.mailSetEntry?._id ?? [mailSet.entries, CUSTOM_MAX_ID]
+				const items = await this.loadMails(lastFetchedId, count)
+				return {
+					items,
+					complete: items.length < count,
+				}
+			},
+
+			loadSingle: (listId, itemId) => this.loadSingleMail([listId, itemId]),
+
+			sortCompare: (mail1, mail2) => sortCompareMailSetMailsReversed(mail1.mail, mail2.mail),
+
+			getItemId: (item) => getElementId(item.mailSetEntry),
+
+			isSameId: (id1, id2) => id1 === id2,
+
+			autoSelectBehavior: () => this.conversationPrefProvider.getMailAutoSelectBehavior(),
+		})
+	}
+
+	get items(): Mail[] {
+		return this.loadedMails().map((mail) => mail.mail)
+	}
+
+	get loadingStatus(): ListLoadingState {
+		return this.listModel.state.loadingStatus
+	}
+
+	get stateStream(): Stream<ListState<Mail>> {
+		return this.listModel.stateStream.map((state) => {
+			const items = state.items.map((item) => item.mail)
+			const selectedItems: Set<Mail> = new Set()
+			for (const item of state.selectedItems) {
+				selectedItems.add(item.mail)
+			}
+			const newState: ListState<Mail> = {
+				...state,
+				items,
+				selectedItems,
+			}
+			return newState
+		})
+	}
+
+	isLoadingAll(): boolean {
+		return this.listModel.state.loadingAll
+	}
+
+	isItemSelected(mailId: Id): boolean {
+		const loadedMail = this.mailMap.get(mailId)
+		if (loadedMail == null) {
+			return false
+		}
+		return this.listModel.isItemSelected(getElementId(loadedMail.mailSetEntry))
+	}
+
+	getMail(mailId: Id): Mail | null {
+		return this.getLoadedMail(mailId)?.mail ?? null
+	}
+
+	getMailSetEntry(mailSetEntryId: Id): MailSetEntry | null {
+		const { mailId } = deconstructMailSetEntryId(mailSetEntryId)
+		return this.getLoadedMail(mailId)?.mailSetEntry ?? null
+	}
+
+	loadAndSelect(mailId: Id, shouldStop: () => boolean): Promise<LoadedMail | null> {
+		return this.listModel.loadAndSelect(mailId, shouldStop)
+	}
+
+	onSingleSelection(mail: Mail) {
+		this.listModel.onSingleSelection(assertNotNull(this.getLoadedMail(mail)))
+	}
+
+	selectNone() {
+		this.listModel.selectNone()
+	}
+
+	cancelLoadAll() {
+		this.listModel.cancelLoadAll()
+	}
+
+	async loadInitial() {
+		await this.listModel.loadInitial()
+	}
+
+	getSelectedAsArray(): Array<Mail> {
+		return this.listModel.getSelectedAsArray().map(({ mail }) => mail)
+	}
+
+	async handleEntityUpdate(update: EntityUpdateData) {
+		if (isUpdateForTypeRef(MailFolderTypeRef, update)) {
```

Comment:
I'm thinking… we could pass labels in also directly now. We should think about it, would be a good point to do it now.